applications/pipeline_visualization/python/pipeline_visualization.py

```diff
@@ -0,0 +1,269 @@
+        # Generate a sine wave signal with time-varying frequency
+        samples = 3000  # Number of samples in the signal
+        duration = 1.0  # Duration of signal in seconds
+        sample_time = duration / samples
+        omega = 2 * np.pi * self.frequency  # Angular frequency
+
+        # Generate the sine wave
+        t = np.arange(samples) * sample_time
+        wave = np.sin(omega * t)
+
+        # Reshape to (samples, 1) to match C++ implementation
+        wave = wave.reshape(samples, 1).astype(np.float32)
+
+        # Gradually increase frequency from 10 to 20 Hz, then wrap back to 10 Hz
+        self.frequency += 0.1
+        if self.frequency > 20.0:
+            self.frequency = 10.0
+
+        # Emit the tensor to the output port
+        op_output.emit(dict(wave=as_tensor(wave)), "out")
+
+
+class ModulateOp(Operator):
+    """Modulation operator that adds high-frequency noise to the input signal.
+
+    This operator receives a time-series signal and adds a 300 Hz sinusoidal modulation
+    with small amplitude (0.05) to simulate measurement noise or signal perturbation.
+    """
+
+    def setup(self, spec: OperatorSpec):
+        spec.input("in")
+        spec.output("out")
+
+    def compute(self, op_input, op_output, context):
+        """Add high-frequency modulation to the input signal."""
+        # Receive the input tensor from the source operator
+        input_tensor = op_input.receive("in").get("wave")
+        samples = input_tensor.shape[0]
+
+        # Parameters for high-frequency modulation/noise
+        frequency = 300.0  # Modulation frequency in Hz
+        amplitude = 0.05  # Amplitude of the modulation
+        duration = 1.0  # Duration of signal in seconds
+        sample_time = duration / samples
+        omega = 2 * np.pi * frequency  # Angular frequency
+
+        # Add modulation to the input signal
+        t = np.arange(samples) * sample_time
+        modulation = amplitude * np.sin(omega * t).reshape(samples, 1)
+        modulated_signal = input_tensor + modulation
+
+        # Ensure output is float32
+        modulated_signal = modulated_signal.astype(np.float32)
+
+        # Emit the modulated signal to the output port
+        op_output.emit(dict(modulated_signal=as_tensor(modulated_signal)), "out")
+
+
+class SinkOp(Operator):
+    """Sink operator that consumes the processed signal.
+
+    This operator acts as the terminal node in the pipeline, receiving the modulated
+    signal. In this example, it simply receives the data without additional processing,
+    but could be extended to perform analysis or visualization.
+    """
+
+    def setup(self, spec: OperatorSpec):
+        spec.input("in")
+
+    def compute(self, op_input, op_output, context):
+        """Receive and process the modulated signal."""
+        # Receive the modulated signal from the previous operator
+        _tensor = next(iter(op_input.receive("in").values()))
```

Comment:
For clarity on the received type, may want to expand to the following equivalent code
```suggestion
        tensormap = op_input.receive("in")
        tensor = next(iter(tensormap.values()))  # get first tensor regardless of name
```